### PR TITLE
chore: refactor error handling to support file range whenever possible.

### PR DESCRIPTION
**Pull request by i4ki**

### CODE REVIEW

File: errors/error.go [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+	Kind Kind
+
+	// Description of the error.
+	Description string
+
+	// FileRange holds the error source.
+	FileRange hcl.Range
+
+	// Stack which originated the error.
+	Stack Stack
```
Comment (katcipis) [inline errors/error.go:42]: hmm why not just use the stack.S type ? then we could create an error related to a stack passing the stack itself, getting a string will allow different parts of the code to pass a different representation of the stack, like some code passing the name, other part passing the absolute path, etc.

File: errors/error.go [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+		case string:
+			e.Description = arg
+		case error:
+			e.Err = arg
+		default:
+			panic(fmt.Errorf("called with unknown type %T", arg))
+		}
+	}
+	if e.isEmpty() {
+		panic(fmt.Errorf("empty error"))
```
Comment (katcipis) [inline errors/error.go:133]:
```suggestion
		panic(errors.New("empty error"))
```

File: errors/error.go [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+		if e.Err != nil {
+			return Is(e.Err, target)
+		}
+		return false
+	}
+
+	// both are *Error
+	if IsKind(e, t.Kind) {
+		return true
+	}
```
Comment (katcipis) [inline errors/error.go:263]: hmm so errors of the same kind but for different stacks would be true when checking errors.Is() ? :thinking: 
we could ignore fields that are not present on the target error, so it would be a partial match, but if both errors have kind + stack and stack differs it seems odd to detect as true, or maybe Im missing something here.

File: errors/error_test.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+			e := errors.E(tc.args...)
+			assert.EqualStrings(t, tc.want, e.Error())
+		})
+	}
+}
+
+func TestErrorIs(t *testing.T) {
+	type testcase struct {
+		name    string
+		args    []interface{}
```
Comment (katcipis) [inline errors/error_test.go:130]:
```suggestion
		err    E
```

Not so sure what we gained with defining test cases like this:

```go
{
			name:   "nil target is not equal",
			args:   []interface{}{"any error"},
			target: nil,
},
```

Instead of (which IMHO reads better):

```go
{
			name:   "nil target is not equal",
			args:   E("any error"),
			target: nil,
},
```

Unless we test when the constructor panics or something, but doesn't seem to be the case here.

File: errors/error.go (hunk 1 above)
Comment (i4ki) [reply errors/error.go:42]: Because `errors` must be imported by every other package, then importing any other TM package here will make cycles...

We can add a `filepath.IsAbs()` check so it helps catch this.

File: errors/error.go (hunk 3 above)
Comment (i4ki) [reply errors/error.go:263]: I thought about this but the same thing could be said about the other metadata/info (filerange, description, etc). Currently (I mean what we have in `main` now) the stack info goes in the error message so we also don't check this. 

We can add this improvement in a separate PR together with the tests changes to make use of it, but I think everything available in the `target` should be used for match. Just don't think make sense adding this now because I'll have to work in the tests to use them, otherwise it's just dead code.

File: errors/error_test.go (hunk 4 above)
Comment (i4ki) [reply errors/error_test.go:130]: yep, this is leftover from the previous iteration.

File: errors/error.go (hunk 1 above)
Comment (katcipis) [reply errors/error.go:42]: As discussed out of the PR, we can also go with an interface to break the cycle, which seems nice IMO, so we can still pass a stack directly when creating errors for it + no need to check things for invariants since the stack itself already did that.

File: cmd/terramate/cli/cli.go [hunk 5]
```diff
@@ -498,7 +497,7 @@ func (c *cli) printStacks() {
 
 	for _, entry := range report.Stacks {
 		stack := entry.Stack
-		stackRepr, ok := c.friendlyFmtDir(stack.PrjAbsPath())
+		stackRepr, ok := c.friendlyFmtDir(stack.Path())
```
[the remark this answers is not in the document]
Comment (i4ki) [reply cmd/terramate/cli/cli.go:500]: those coverage warnings are not correct... these lines are tested with e2e test...

File: errors/error.go [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+	// Description of the error.
+	Description string
+
+	// FileRange holds the error source.
+	FileRange hcl.Range
+
+	// Stack which originated the error.
+	Stack StackMeta
+
+	// Err represents the underlying error.
```
Comment (katcipis) [inline errors/error.go:44]:
```suggestion
	// Err holds the underlying error.
```

File: errors/error.go [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+	Err error
+}
+
+type (
+	// Kind defines the kind of error.
```
Comment (katcipis) [inline errors/error.go:49]:
```suggestion
	// Kind defines the kind of an error.
```

File: errors/error.go [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+	Kind string
+	// StackMeta has the metadata of the stack which originated the error.
+	// Same interface as stack.Metadata.
```
Comment (katcipis) [inline errors/error.go:52]:
```suggestion
	// StackMeta provides metadata about the stack that originated the error.
```

thinking if it makes sense to explicit document that the interface is copied :thinking: 

Also thinking if StackMetadata would not be more clear (could be just me more used to always referencing it as metadata not only as meta).

File: errors/error.go [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+// The type of each argument determines its meaning. If more than one argument
+// of a given type is presented, only the last one is recorded.
+//
+// The types are:
+// 	errors.Kind
+//		The kind of error (eg.: HCLSyntax, TerramateSchema, etc).
+//	hcl.Range
+//		The file range where the error originated.
+//  errors.Stack
+//		The stack which the error originated.
```
Comment (katcipis) [inline errors/error.go:73]:
```suggestion
//		errors.StackMeta
//		The stack that originated the error.
```

File: errors/error.go [hunk 10]
```diff
@@ -0,0 +1,373 @@
...
+//	error
+//		The underlying error that triggered this one.
+//	hcl.Diagnostics
+//		The underlying hcl error that triggered this one.
+//		If this error's FileRange is not set, the diagnostic subject range is
```
Comment (katcipis) [inline errors/error.go:78]:
```suggestion
//		If hcl.Range is not set, the diagnostic subject range is
```

File: errors/error.go [hunk 11]
```diff
@@ -0,0 +1,373 @@
...
+//		pulled.
+//		If this error's Description is not set, the diagnostic detail field is
```
Comment (katcipis) [inline errors/error.go:80]:
```suggestion
//		If the string Description is not set, the diagnostic detail field is
```

File: errors/error.go [hunk 12]
```diff
@@ -0,0 +1,373 @@
...
+//		pulled.
+//	string
+//		The error description. It supports formatting using the Go's fmt verbs
+//      as long as the arguments are not of the defined types.
```
Comment (katcipis) [inline errors/error.go:84]:
```suggestion
//	string
//		The error description. It supports formatting using the Go's fmt verbs
//		as long as the arguments are not one of the defined types.
```

File: errors/error.go [hunk 13]
```diff
@@ -0,0 +1,373 @@
...
+//
+// If the error is printed, only those items that have been
+// set to non-zero values will appear in the result. For the `hcl.Range` type,
+// the `range.Empty()` method is used.
+//
+// Error promotions:
+//
+// If Kind is not specified or Any, we set it to the Kind of
```
Comment (katcipis) [inline errors/error.go:92]: Do we have an  "Any" kind ?

File: errors/error.go [hunk 14]
```diff
@@ -0,0 +1,373 @@
...
+// the underlying error.
+// If FileRange is not specified, we set it to the FileRange of the underlying
+// error (if set).
```
Comment (katcipis) [inline errors/error.go:95]:
```suggestion
// The following fields are promoted from underlying errors when absent:
//
// - errors.Kind
// - errors.StackMeta
// - hcl.Range
```

File: errors/error.go [hunk 15]
```diff
@@ -0,0 +1,373 @@
...
+//
+// Minimization:
+//
+// In order to avoid duplicated messages, we erase the fields present in the
+// underlying error if already set with same value in this error.
```
Comment (katcipis) [inline errors/error.go:100]:
```suggestion
// In order to avoid duplicated messages, when wrapping an errors.E inside another errors.E 
// any fields present on the current error overrides the fields defined on the wrapped error.
```

File: errors/error.go [hunk 16]
```diff
@@ -0,0 +1,373 @@
...
+		case Kind:
+			e.Kind = arg
+		case hcl.Range:
+			e.FileRange = arg
+		case hcl.Diagnostics:
+			diags = arg
+			if diags.HasErrors() {
+				diag := diags[0]
+				if diag.Subject != nil {
+					e.FileRange = *diag.Subject
```
Comment (katcipis) [inline errors/error.go:125]: The range...on the diagnostics...is called SUBJECT ?

terror

File: errors/error.go [hunk 17]
```diff
@@ -0,0 +1,373 @@
...
+}
+
+// Is tells if err (or any of its underlying errors) matches target.
+// It works with any error but if comparing *Error type it uses the Kind field,
+// otherwise it fallback to standard errors.Is().
+func Is(err, target error) bool {
+	if (err == nil) != (target == nil) {
+		return false
+	}
+	e, ok := err.(*Error)
```
Comment (katcipis) [inline errors/error.go:303]: when (err == nil) && (target == nil) this will panic, do we want it to panic ? Go's errors.Is returns true if both errors are nil: https://cs.opensource.google/go/go/+/refs/tags/go1.18.1:src/errors/wrap.go;l=41

File: errors/error.go [hunk 18]
```diff
@@ -0,0 +1,373 @@
...
+	}
+	if e.Kind != "" && e.Kind == k {
+		return true
+	}
+	return IsKind(e.Err, k)
+}
+
+// Is tells if err (or any of its underlying errors) matches target.
+// It works with any error but if comparing *Error type it uses the Kind field,
+// otherwise it fallback to standard errors.Is().
```
Comment (katcipis) [inline errors/error.go:298]:
```suggestion
// It works with any error but when comparing with an errors.Error type it will try to match the following fields:
// - Kind
// - Description
// - Stack
// - FileRange
// - Err
// Any fields absent (empty) on the target error are ignored even if they exist on err (partial match).
// If the target error is no an error.Error it fallbacks to standard errors.Is().
```

File: errors/error.go [hunk 19]
```diff
@@ -0,0 +1,373 @@
...
+		return true
+	}
+	if e.Err != nil {
+		return hasDesc(e.Err, desc)
+	}
+	return false
+}
+
+// hasSameStack recursively check if err or any of its underlying errors has the
+// provided stack set.
```
Comment (katcipis) [inline errors/error.go:358]:
```suggestion
// hasSameStack recursively check if err or any of its underlying errors has
// the same stack metadata.
```

**Review by katcipis — COMMENTED** [review]
Some more comments but still haven't finished it :-)

File: errors/error.go (hunk 10 above)
Comment (katcipis) [reply errors/error.go:78]: Since on the context of the errors.E constructor we are usually talking about the types of the variadic args passed, not the field on the struct they will be stored.

File: errors/error.go [hunk 20]
```diff
@@ -0,0 +1,373 @@
...
+	emptyRange := hcl.Range{}
+	if e.FileRange == emptyRange {
+		e.FileRange = prev.FileRange
+	}
+	if prev.FileRange == e.FileRange {
+		prev.FileRange = hcl.Range{}
+	}
+	if prev.Stack == e.Stack {
+		prev.Stack = nil
+	}
```
[the remark this answers is not in the document]
Comment (katcipis) [reply errors/error.go:182]: Also, feel free to disregard this because maybe Im being pedantic, but comparing using interfaces here seems risky if there is any other field on the underlying object that implements the interface that may differ and we are not interested in. For our purposes we are interested on then having the same name/path/description and that is it. Also, even though it is a stretch, if there is any type implementing this interface that is not comparable this would panic.

When doing the error matching we are already comparing by calling the interface methods: https://github.com/mineiros-io/terramate/pull/291/files#diff-d128886dfda28afbfb957c032a184b95f625ca12ff25c8656247783f3e9bf112R364

So maybe we could do it here too.

File: errors/error_test.go [hunk 21]
```diff
@@ -0,0 +1,217 @@
...
+	"testing"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/madlambda/spells/assert"
+	"github.com/mineiros-io/terramate/errors"
+)
+
+var E = errors.E
+
+var (
```
Comment (katcipis) [inline errors/error_test.go:30]:
```suggestion
const (
```

File: errors/error_test.go [hunk 22]
```diff
@@ -0,0 +1,217 @@
...
+)
+
+func TestNoArgs(t *testing.T) {
+	defer func() {
+		err := recover()
+		if err == nil {
+			t.Fatal("E() did not panic")
+		}
+	}()
+	_ = E()
```
Comment (katcipis) [inline errors/error_test.go:43]: Maybe we could have something like a assert.Panic(func() {}) that executes the given function and checks it panics.

File: errors/error_test.go [hunk 23]
```diff
@@ -0,0 +1,217 @@
...
+		{
+			name: "nested errors with last one with range",
+			err: E("1", E("2", E("3", hcl.Range{
+				Filename: "test.tm",
+				Start:    hcl.Pos{Line: 1, Column: 5, Byte: 3},
+				End:      hcl.Pos{Line: 1, Column: 10, Byte: 13},
+			})),
+			),
+			want: "test.tm:1,5-10: 1: 2: 3",
+		},
```
Comment (katcipis) [inline errors/error_test.go:141]: I think we are missing the behavior when building errors using hcl.Diagnostics here.

File: errors/error.go [hunk 24]
```diff
@@ -0,0 +1,373 @@
...
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/hcl/v2"
+)
+
+// Error is the default Terramate error type.
+// It may contains some fields unset, but not all.
```
[the remark this answers is not in the document]
Comment (i4ki) [reply errors/error.go:29]: I know "partial" means incompletion but saying it could be partially set imply at least one must be set?

What about: At least one of the error fields must be set.
?

File: errors/error_test.go [hunk 25]
```diff
@@ -0,0 +1,217 @@
...
+
+func TestErrorIs(t *testing.T) {
+	type testcase struct {
+		name    string
+		err     error
+		target  error
+		areSame bool
+	}
+
+	for _, tc := range []testcase{
```
Comment (katcipis) [inline errors/error_test.go:157]: Maybe we could add a test to define when both are nil here.

File: errors/testing.go [hunk 26]
```diff
@@ -0,0 +1,62 @@
...
+	t.Helper()
+	if (got == nil) != (want == nil) {
+		t.Fatalf("got error[%v] differs from want[%v]", got, want)
+	}
+	if want == nil {
+		return
+	}
+	e1, ok := got.(*Error)
+	if !ok {
+		t.Fatal("got is not an *errors.Error")
```
Comment (katcipis) [inline errors/testing.go:30]:
```suggestion
		t.Fatalf("got %v is not an *errors.Error", got)
```

File: errors/testing.go [hunk 27]
```diff
@@ -0,0 +1,62 @@
...
+	}
+
+	e2, ok := want.(*Error)
+	if !ok {
+		t.Fatal("want is not an *errors.Error")
```
Comment (katcipis) [inline errors/testing.go:35]:
```suggestion
		t.Fatalf("want %v is not an *errors.Error", want)
```

File: errors/testing.go [hunk 28]
```diff
@@ -0,0 +1,62 @@
...
+// AssertIsKind asserts err is of kind k.
+func AssertIsKind(t *testing.T, err error, k Kind) {
+	t.Helper()
+	if !IsKind(err, k) {
+		t.Fatalf("error[%v] is not of kind %q", err, k)
+	}
+}
+
+// Assert err is (contains, wraps, etc) target.
+func Assert(t *testing.T, err, target error) {
```
Comment (katcipis) [inline errors/testing.go:50]:
```suggestion
// AssertIs err is (contains, wraps, etc) target.
func AssertIs(t *testing.T, err, target error) {
```

I think just assert is too vague, specially since this is a specialized assertion built on top of "is" behavior (not some simple boolean check etc).

File: errors/error.go (hunk 8 above)
Comment (i4ki) [reply errors/error.go:52]: I think it's important to add this hint because changes to this type must be aligned with the definition of stack metadata in the stack package.
Now thinking about this I also would say that we should add a similar hint in the stack package like:

```
// Metadata has all metadata loaded per stack
// Hint: changes to this interface must be replicated in errors.StackMeta.
```

What do you think?

File: errors/testing.go [hunk 29]
```diff
@@ -0,0 +1,62 @@
+// Copyright 2022 Mineiros GmbH
```
Comment (katcipis) [inline errors/testing.go:1]: The idea of exporting some test facilities is really cool, Im just not so sure if we should do it in the same package/namespace of the actual code, like when you see the package docs there will also be a bunch of assertion functions there...maybe it is OK and it is just some subjective bias because I never saw it before, but could make it more annoying for people who are just looking to use the errors API and not actually write tests (although they should write tests too :laughing: ).

Maybe we could go with "test/errors" like we have for other testing facilities on the project.

**Review by katcipis — COMMENTED** [review]
Not finished yet but GH interface is dumb and doesn't allow me to answer any discussions while Im reviewing code....because....

File: errors/error.go (hunk 8 above)
Comment (katcipis) [reply errors/error.go:52]: Hmm I disagree. Any incompatibilties on the interfaces will be caught at build time already, IMO it only clobbers the docs (imagine a third interface somewhere else, now we have to add comments on all places about all other places). But if you want to do it Im OK, not a strong opinion or anything.

File: errors/error.go (hunk 13 above)
Comment (i4ki) [reply errors/error.go:92]: outdated docs, thanks!

File: errors/error.go (hunk 14 above)
Comment (katcipis) [reply errors/error.go]: I thought intuitively that stacks were going to be promoted, but I think I was wrong on that so disregard this

File: generate/generate.go [hunk 30]
```diff
@@ -73,20 +72,19 @@ func Do(root string, workingDir string) Report {
 		stack stack.S,
 		globals terramate.Globals,
 	) stackReport {
-		stackpath := stack.AbsPath()
+		stackpath := stack.HostPath()
```
Comment (katcipis) [inline generate/generate.go:75]: Noice

File: errors/error.go (hunk 15 above)
Comment (i4ki) [reply errors/error.go:99]: overrides is not the correct term because if the underlying error has different field values, we keep them for the error context. We just clear the fields in the underlying error *when* they have the same value in the current error.
I also don't know if saying "wrapping" is a good term because this concept in Go already means wrapping errors with `%w` and this is not supported here.

File: globals_test.go [hunk 31]
```diff
@@ -1067,21 +1068,28 @@ func TestLoadGlobals(t *testing.T) {
...
 			wantGlobals := tcase.want
 
-			stacks := s.LoadStacks()
-			for _, stack := range stacks {
-				stackMeta := stack.Meta()
-				got, err := terramate.LoadStackGlobals(s.RootDir(), stackMeta)
+			stackEntries, err := terramate.ListStacks(s.RootDir())
+			if errors.IsKind(tcase.wantErr, hcl.ErrTerramateSchema) {
+				errors.AssertKind(t, err, tcase.wantErr)
+			}
```
Comment (katcipis) [inline globals_test.go:1074]: When tcase.wantErr != hcl.ErrTerramateSchema (like nil) is it safe/best to go on with the test if listing stacks failed ? I imagine the stackEntries will be an empty slice and the test failure will be harder to understand because there will be no indication that stack listing failed (the error will indicate that something wrong happened loading the globals) :thinking:

File: errors/error.go (hunk 16 above)
Comment (i4ki) [reply errors/error.go:124]: it kinda makes sense for me... because just Range says nothing, FileRange says a bit more but still what range? The word *Subject* at least points that this field contains the offending part, the topic of the diagnostics, and so on. They also provide a *Context* field with a broader range so you can print/render this part of the code in an error in order to easily spot the problem because sometimes the `Subject` is a tiny expression inside an expression and so on.
But Context can be nil, which means the Context == Subject, but they don't provide any helper method for that. For the extension the tighter range is enough but I think we can provide better error messages using the Context (if we get rid of logging for showing errors in text mode).

Comment (katcipis) [reply errors/error.go:124]: was unaware they had two ranges, if there is just one it is simple to me that it is the file range of the source that caused the problem/diagnostic and calling subject is just confusing, but if they have multiple ranges then it may make sense.

File: errors/error.go (hunk 8 above)
Comment (i4ki) [reply errors/error.go:52]: > Any incompatibilties on the interfaces will be caught at build time already
how? errors.E() receives interface{}...

Comment (katcipis) [reply errors/error.go:52]: ah true.. still it seems odd to keep documenting this and hoping it will be enough to avoid issues, for a scenario like that it would be safer to just define a single interface/type and use it on all places then... but that would introduce some sort of types package.

File: errors/error.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (i4ki) [reply errors/error.go:180]: this comparison is there since errors.Stack was a string. It needs to be fixed. Thanks for catching it.

File: hcl/hcl.go [hunk 32]
```diff
@@ -92,22 +97,108 @@ type Stack struct {
 // Blocks maps a filename to a slice of HCL blocks associated with it
 type Blocks map[string][]*hclsyntax.Block
 
-// Errors returned when parsing Terramate configuration.
-const (
-	ErrHCLSyntax                errutil.Error = "HCL syntax error"
-	ErrMalformedTerramateConfig errutil.Error = "malformed terramate config"
-	ErrMalformedTerraform       errutil.Error = "malformed terraform"
-)
+// TerramateParser is an HCL parser tailored for Terramate configurion schema.
```
Comment (katcipis) [inline hcl/hcl.go:100]:
```suggestion
// TerramateParser is an HCL parser tailored for Terramate configuration schema.
```

File: errors/error.go (hunk 19 above)
Comment (i4ki) [reply errors/error.go:362]: this message would be accurate if both params were errors and we were looking if they have the same metadata.

Comment (katcipis) [reply errors/error.go:362]: hmm but what would be a "stack set" here ? I just don't get it :thinking:

Comment (i4ki) [reply errors/error.go:362]: ... has the provided stack set to its Stack field. ?

Comment (katcipis) [reply errors/error.go:362]: hmm for some reason I was always thinking about "sets"...like a set of stacks :laughing: . Maybe Im just extra stupid..but I was trying to improve this because it is confusing at least to me. But it is probably an isolated dumbness on my part, disregard.

File: hcl/hcl.go [hunk 33]
```diff
@@ -173,45 +264,36 @@ func ParseModules(path string) ([]Module, error) {
...
 		if block.Type != "module" {
 			continue
 		}
 
 		if len(block.Labels) != 1 {
-			return nil, errutil.Chain(
-				ErrMalformedTerraform,
-				fmt.Errorf("module block must have 1 label"),
-			)
+			return nil, errors.E(ErrTerraformSchema, block.OpenBraceRange,
```
Comment (katcipis) [inline hcl/hcl.go:281]: nice :rocket:

File: hcl/hcl_test.go [hunk 34]
```diff
@@ -132,30 +134,38 @@ module "bleh" {
...
+			configdir := t.TempDir()
+			tfpath := test.WriteFile(t, configdir, "main.tf", tc.input)
 
-			modules, err := hcl.ParseModules(path)
-			assert.IsError(t, err, tc.want.err)
+			if tc.want.err != nil {
+				if e, ok := tc.want.err.(*errors.Error); ok {
+					if e.FileRange.Filename != "" {
+						e.FileRange.Filename = tfpath
+					}
```
Comment (katcipis) [inline hcl/hcl_test.go:163]: hmm if the wanted error defines a filename in the range we overwrite it, if none is defined (empty string) we leave it empty. Is this the behavior we want here ? (usually we set automatically something when none is defined, like a default)

File: hcl/hcl_test.go [hunk 35]
```diff
@@ -1009,6 +1044,24 @@ func testParser(t *testing.T, tc testcase) {
 	}
 }
 
+// some helpers to easy build file ranges.
```
Comment (katcipis) [inline hcl/hcl_test.go:1047]: noice :rocket:

File: init.go [hunk 36]
```diff
@@ -39,47 +40,49 @@ func Init(root, dir string) error {
...
+		return errors.E(errInit, err)
 	}
 
 	if found {
-		return fmt.Errorf("directory %q is inside stack %q but nested stacks are disallowed",
-			dir, parentStack.PrjAbsPath())
+		return errors.E(errInit,
+			"directory %q is inside stack %q but nested stacks are disallowed",
+			dir, parentStack.Path(),
+		)
```
Comment (katcipis) [inline init.go:78]:
```suggestion
		return errors.E(errInit, parentStack,
			"directory %q is inside stack but nested stacks are disallowed",
			dir,
		)
```

File: list.go [hunk 37]
```diff
@@ -50,18 +49,14 @@ func ListStacks(root string) ([]Entry, error) {
 				return filepath.SkipDir
 			}
 
-			logger.Trace().
-				Str("stack", path).
-				Msg("Try load stack.")
+			logger.Trace().Str("stack", path).Msg("Try load stack.")
```
Comment (katcipis) [inline list.go:52]: noice :100:

**Review by katcipis — APPROVED** [review]
LGTM + bunch of comments where Im possibly just wrong/being dumb so feel free to disregard.

File: errors/error_test.go (hunk 22 above)
Comment (i4ki) [reply errors/error_test.go:43]: cool

File: errors/error.go (hunk 17 above)
Comment (i4ki) [reply errors/error.go:302]: fixed.

File: errors/error_test.go (hunk 23 above)
Comment (i4ki) [reply errors/error_test.go:141]: fixed

File: errors/error_test.go (hunk 25 above)
Comment (i4ki) [reply errors/error_test.go:179]: fixed

File: errors/testing.go (hunk 29 above)
Comment (i4ki) [reply errors/testing.go:1]: yeah I thought the same thing but I didn't know it was worth the effort. Will do.

File: hcl/hcl_test.go (hunk 34 above)
Comment (i4ki) [reply hcl/hcl_test.go]: fixed

File: globals_test.go (hunk 31 above)
Comment (i4ki) [reply globals_test.go]: fixed

File: errors/testing.go (hunk 28 above)
Comment (i4ki) [reply errors/testing.go]: the context is in the package name: `errors.Assert(err1, err2)`.. it assert that errors match.

Comment (katcipis) [reply errors/testing.go]: hmm not so sure that is clear enough, assertions are usually about something being true (predicates) or comparing things (equality). The matching logic implemented by errors.Is (Go's and ours) is different from most classical assertions. One example is that this assertion function is not symmetrical, `errors.Assert(t, err1, err2)` != `errors.Assert(t, err2, err1)`, for a function called just Assert, even though inside the errors package, that is a little odd IMHO. But not a huge deal, just sharing my perception on this.

**Review by katcipis — APPROVED** [review]
YEAHHH :rocket:
